### PR TITLE
[download-bundle] Update mono bundle download URL

### PR DESCRIPTION
**Pull request by jonpryor**
Around 2018-Aug-15, the Jenkins config for xamarin-android was changed
in a way which [altered the URL for where files are uploaded][0] so
that it contained `${BUILD_NUMBER}`, which is a value which increments
on every build.  A consequence of this change is that the bundle could
no longer used, as it could no longer be *found*, as there is no way
of knowing what `${BUILD_NUMBER}` could possibly be, which in turn
caused increased build times for PRs and master builds, as mono would
now always need to be rebuilt.

On 2018-Sep-7, the [Jenkins config was updated][1] to remove
`${BUILD_NUMBER}` from the URL, allowing for "static" URLs which works
with our bundle mechansism.

With the removal of `${BUILD_NUMBER}`, we're good, right?  The bundle
should now be used?

Not so much.  One other part needs updating: the *rest* of the URL
*also* changed as part of the 2018-Aug-15 change!
`download-bundle.targets` was attempting to download e.g.:

	https://xamjenkinsartifact.azureedge.net/xamarin-android/xamarin-android/bin/Debug/bundle-v21-h7419ab9f-Debug-Darwin-libzip=b95cf3f,llvm=bdb3a11,mono=cca6805.zip

However, the *correct* URL we now need to download is e.g.:

	https://xamjenkinsartifact.blob.core.windows.net/mono-jenkins/xamarin-android/bin/Debug/bundle-v21-h7419ab9f-Debug-Darwin-libzip=b95cf3f,llvm=bdb3a11,mono=cca6805.zip

Update `$(_AzureBaseUri)` so that we now download from the correct
base URL, so that the mono bundle can actually be downloaded.

[0]: https://jenkins.mono-project.com/view/Xamarin.Android/job/xamarin-android/jobConfigHistory/showDiffFiles?timestamp1=2018-07-23_19-09-25&timestamp2=2018-08-15_22-31-28
[1]: https://jenkins.mono-project.com/view/Xamarin.Android/job/xamarin-android/jobConfigHistory/showDiffFiles?timestamp1=2018-08-31_14-49-55&timestamp2=2018-09-07_18-49-53

### CODE REVIEW

**Comment by jonpryor** [comment]
***NOTE***: *Before* merging, *verify* the fix by downloading the **View as plain text** link of the PR build, and searching for `_DownloadBundle:`, and ensure it *actually* succeeded.

For example, consider [this PR build](https://jenkins.mono-project.com/job/xamarin-android-pr-builder/3961/) from PR #2163, which contains:

```
_DownloadBundle:
  Downloading `https://****.azureedge.net/xamarin-android/xamarin-android/bin/Debug/bundle-v21-h7419ab9f-Debug-Darwin-libzip=b95cf3f,llvm=bdb3a11,mono=cca6805.zip` to `/Users/builder/android-archives/.bundle-v21-h7419ab9f-Debug-Darwin-libzip=b95cf3f,llvm=bdb3a11,mono=cca6805.zip.download`.
/Users/builder/jenkins/workspace/xamarin-android-pr-builder/xamarin-android/build-tools/download-bundle/download-bundle.targets(24,5): warning : Unable to download URL `https://****.azureedge.net/xamarin-android/xamarin-android/bin/Debug/bundle-v21-h7419ab9f-Debug-Darwin-libzip=b95cf3f,llvm=bdb3a11,mono=cca6805.zip` to `/Users/builder/android-archives/bundle-v21-h7419ab9f-Debug-Darwin-libzip=b95cf3f,llvm=bdb3a11,mono=cca6805.zip`: 404 (Not Found) [/Users/builder/jenkins/workspace/xamarin-android-pr-builder/xamarin-android/build-tools/download-bundle/download-bundle.csproj]
  The previous error was converted to a warning because the task was called with ContinueOnError=true.
/Users/builder/jenkins/workspace/xamarin-android-pr-builder/xamarin-android/build-tools/download-bundle/download-bundle.targets(24,5): warning : 404 (Not Found) [/Users/builder/jenkins/workspace/xamarin-android-pr-builder/xamarin-android/build-tools/download-bundle/download-bundle.csproj]
```

`404 (Not Found)` is *not* what we want to see. :-)

**Comment by jonpryor** [comment]
Fortunately, we don't need a full build to verify this fix: the [**Download as plain text** link](https://jenkins.mono-project.com/job/xamarin-android-pr-builder/3962/consoleText) for this PR contains:

```
_DownloadBundle:
  Downloading `https://****.blob.core.windows.net/mono-jenkins/xamarin-android/bin/Debug/bundle-v21-h7419ab9f-Debug-Darwin-libzip=b95cf3f,llvm=bdb3a11,mono=cca6805.zip` to `/Users/builder/android-archives/.bundle-v21-h7419ab9f-Debug-Darwin-libzip=b95cf3f,llvm=bdb3a11,mono=cca6805.zip.download`.
Archive:  /Users/builder/android-archives/bundle-v21-h7419ab9f-Debug-Darwin-libzip=b95cf3f,llvm=bdb3a11,mono=cca6805.zip
  inflating: /var/folders/19/hjm21bn17913_0ck7nh3thdr0000gn/T/08t1g5yp.uqa/8jckhnzo.8g7/lib/xamarin.android/xbuild-frameworks/MonoAndroid/v1.0/I18N.CJK.dll  
  inflating: /var/folders/19/hjm21bn17913_0ck7nh3thdr0000gn/T/08t1g5yp.uqa/8jckhnzo.8g7/lib/xamarin.android/xbuild-frameworks/MonoAndroid/v1.0/I18N.dll  
...
```

Success!

**Comment by jonpryor** [comment]
@jonathanpeppers: done.

**Comment by jonpryor** [comment]
One "curious" thing is that even though it now downloads the bundle...the build times aren't significantly different.  PR #2163 built in 2h10min, while this PR built in 2h5min.

If the bundle isn't being used, why is the build time so similar?

The PR #2163 build log `msbuild-20180908T030426-prepare-msbuild.binlog` contains:

```
Target "_BuildUnlessCached" in file "/Users/builder/jenkins/workspace/xamarin-android-pr-builder/xamarin-android/src/mono-runtimes/mono-runtimes.targets" from project "/Users/builder/jenkins/workspace/xamarin-android-pr-builder/xamarin-android/src/mono-runtimes/mono-runtimes.csproj" (target "Build" depends on it):
Building target "_BuildUnlessCached" completely.
...
Target "_Autogen" in file "/Users/builder/jenkins/workspace/xamarin-android-pr-builder/xamarin-android/src/mono-runtimes/mono-runtimes.targets" from project "/Users/builder/jenkins/workspace/xamarin-android-pr-builder/xamarin-android/src/mono-runtimes/mono-runtimes.csproj" (target "ForceBuild" depends on it):
...
Target "_BuildRuntimes" in file "/Users/builder/jenkins/workspace/xamarin-android-pr-builder/xamarin-android/src/mono-runtimes/mono-runtimes.targets" from project "/Users/builder/jenkins/workspace/xamarin-android-pr-builder/xamarin-android/src/mono-runtimes/mono-runtimes.csproj" (target "ForceBuild" depends on it):
Building target "_BuildRuntimes" completely.
No input files were specified.
Task "Exec"
  Task Parameter:Command=make DISABLE_IOS=1 -j4 package-android-armeabi-v7a package-android-x86 package-android-host-Darwin  CONFIGURATION=debug IGNORE_PROVISION_MXE=true IGNORE_PROVISION_ANDROID=true IGNORE_PACKAGE_LLVM=true ANDROID_TOOLCHAIN_DIR="/Users/builder/android-toolchain" ANDROID_TOOLCHAIN_CACHE_DIR="/Users/builder/android-archives" ANDROID_TOOLCHAIN_PREFIX="/Users/builder/android-toolchain/toolchains" LLVM_SRC="/Users/builder/jenkins/workspace/xamarin-android-pr-builder/xamarin-android/src/mono-runtimes/../../external/llvm" MXE_PREFIX_DIR="/Users/builder/android-toolchain" MXE_SRC="/Users/builder/jenkins/workspace/xamarin-android-pr-builder/xamarin-android/src/mono-runtimes/../../external/mxe"
```

So it *is* asking `mono` to build...

Use *this* command:

```bash
$ msbuild /consoleloggerparameters:PerformanceSummary msbuild-20180908T030426-prepare-msbuild.binlog > x.txt
```

And `x.txt` states:

```
Target Performance Summary:
    ...
    77652 ms  _BuildRuntimes                             1 calls
    81216 ms  _BuildUnlessCached                         3 calls
```

Building mono took 77s, which is surprisingly fast for a mono build...

...because I'm looking at the wrong log file.  Look at `msbuild-20180908T024945-prepare-msbuild.binlog`, and it's *much* longer:

```
Target Performance Summary:
    ...
    811440 ms  _BuildRuntimes                             1 calls
```

Or ~13.5 minutes.  Which *still* doesn't explain a ~5 minute perf difference.

In short, as per build log [#3496](https://jenkins.mono-project.com/job/xamarin-android-pr-builder/3962/) things are improved because the bundle is actually used, but the improvement isn't as much as I'd expect, *especially* after digging in further.  (Now I'm just more confused.)

**Comment by radekdoulik** [comment]
There's ~11.5 minutes difference in running all the tests, which is supposedly unrelated. Makes one wonder about such big test times difference though.

```
    11132021 ms  /Users/builder/jenkins/workspace/xamarin-android-pr-builder/xamarin-android/build-tools/scripts/RunTests.targets   5 calls
```
versus
```
    11829098 ms  /Users/builder/jenkins/workspace/xamarin-android-pr-builder/xamarin-android/build-tools/scripts/RunTests.targets   5 calls
```

**Comment by jonpryor** [comment]
lol, attempting to change the `$(_AzureBaseUri)` value had no effect on the PR builder, because [the file had *already* been downloaded](https://jenkins.mono-project.com/job/xamarin-android-pr-builder/3964/):

```
Skipping target "_DownloadBundle" because all output files are up-to-date with respect to the input files.
```
